plugins/renepay/flow.c

```diff
@@ -0,0 +1,849 @@
+			__LINE__,
+			type_to_string(this_ctx, struct amount_msat, &B),
+			type_to_string(this_ctx, struct amount_msat, &A));
+	}
+	struct amount_msat numerator; // MAX(0,B-f)
+
+	if(!amount_msat_sub(&numerator,B,f))
+		numerator = AMOUNT_MSAT(0);
+
+	tal_free(this_ctx);
+	return amount_msat_less_eq(f,A) ? 1.0 : amount_msat_ratio(numerator,denominator);
+}
+
+
+
+
+
+
+// TODO(eduardo): remove this function, is a duplicate
+void remove_completed_flow(const struct gossmap *gossmap,
+			   struct chan_extra_map *chan_extra_map,
+			   struct flow *flow)
+{
+	for (size_t i = 0; i < tal_count(flow->path); i++) {
+		struct chan_extra_half *h = get_chan_extra_half_by_chan(gossmap,
+							       chan_extra_map,
+							       flow->path[i],
+							       flow->dirs[i]);
+		if (!amount_msat_sub(&h->htlc_total, h->htlc_total, flow->amounts[i]))
+		{
+			debug_err("%s could not substract HTLC amounts, "
+				   "half total htlc amount = %s, "
+				   "flow->amounts[%lld] = %s.",
+				   __PRETTY_FUNCTION__,
+				   type_to_string(tmpctx, struct amount_msat, &h->htlc_total),
+				   i,
+				   type_to_string(tmpctx, struct amount_msat, &flow->amounts[i]));
+		}
+		if (h->num_htlcs == 0)
+		{
+			debug_err("%s could not decrease HTLC count.",
+				   __PRETTY_FUNCTION__);
+		}
+		h->num_htlcs--;
+	}
+}
+// TODO(eduardo): remove this function, is a duplicate
+void remove_completed_flow_set(
+		const struct gossmap *gossmap,
+		struct chan_extra_map *chan_extra_map,
+		struct flow **flows)
+{
+	for(size_t i=0;i<tal_count(flows);++i)
+	{
+		remove_completed_flow(gossmap,chan_extra_map,flows[i]);
+	}
+}
+
+// TODO(eduardo): remove this function, is a duplicate
+void commit_flow(
+		const struct gossmap *gossmap,
+		struct chan_extra_map *chan_extra_map,
+		struct flow *flow)
+{
+	for (size_t i = 0; i < tal_count(flow->path); i++) {
+		struct chan_extra_half *h = get_chan_extra_half_by_chan(gossmap,
+							       chan_extra_map,
+							       flow->path[i],
+							       flow->dirs[i]);
+		if (!amount_msat_add(&h->htlc_total, h->htlc_total, flow->amounts[i]))
+		{
+			debug_err("%s could not add HTLC amounts, "
+				   "flow->amounts[%lld] = %s.",
+				   __PRETTY_FUNCTION__,
+				   i,
+				   type_to_string(tmpctx, struct amount_msat, &flow->amounts[i]));
+		}
+		h->num_htlcs++;
+	}
+}
+// TODO(eduardo): remove this function, is a duplicate
+void commit_flow_set(
+		const struct gossmap *gossmap,
+		struct chan_extra_map *chan_extra_map,
+		struct flow **flows)
+{
+	for(size_t i=0;i<tal_count(flows);++i)
+	{
+		commit_flow(gossmap,chan_extra_map,flows[i]);
+	}
+}
+
+/* Helper function to fill in amounts and success_prob for flow
+ *
+ * IMPORTANT: here we do not commit flows to chan_extra, flows are commited
+ * after we send those htlc.
+ *
+ * IMPORTANT: flow->success_prob is misleading, because that's the prob. of
+ * success provided that there are no other flows in the current MPP flow set.
+ * */
+void flow_complete(struct flow *flow,
```

Comment:
ok I see here one computes statistics after the flow has been computed. Is that correct?

Reply:
We use that function to compute the real fees along the path and the prob of success of that flow.
There are two aspects here:
1. MCF does not compute fees exactly because the standard MCF algorithms assume a flow conservation, while in reality msats are leaked as routing fees along the path. `flow complete` takes a flow we have already solved with the MCF algorithm and computes the real fees that we need later to inform `sendpay`.
2. `flow_complete` also estimates a probability of success for that flow, however, that probability assumes that there are no other flows in this same payment taken liquidity from this path. 
By multiplying the prob. of success of every flow in the payment one can estimate the prob. of success of that payment. But in reality this number tends to be smaller than that.

There is another function called `flow_set_probability` that computes the probability of success of a given bundle of flows.

Reply:
In the payment notes (`renepaystatus`) and logs you can see now for every attempt the fees, delay, amount and probability of success (conditional). https://github.com/ElementsProject/lightning/pull/6376/commits/f2a65458b2eefd314c78302f640824244dcfa89a